src/Xamarin.Android.Build.Tasks/Utilities/TypeMappingReleaseNativeAssemblyGenerator.cs

```diff
@@ -208,73 +213,152 @@ void InitMapModules ()
 			}
 
-			var sortedTokens = tokens.Keys.ToArray ();
-			Array.Sort (sortedTokens);
+			mapModuleEntries.Sort ((StructureInstance<TypeMapModuleEntry> a, StructureInstance<TypeMapModuleEntry> b) => a.Obj.type_token_id.CompareTo (b.Obj.type_token_id));
+			allModulesData.Add (new ModuleMapData (moduleDataSymbolLabel, mapModuleEntries));
 
-			var moduleData = new ModuleMapData (moduleDataSymbolLabel);
-			foreach (uint token in sortedTokens) {
-				var map_entry = new TypeMapModuleEntry {
-					type_token_id = token,
-					java_map_index = tokens[token],
-				};
+			uint GetJavaEntryIndex (string javaTypeName)
+			{
+				if (!javaTypesByName.TryGetValue (javaTypeName, out TypeMapJava javaType)) {
+					throw new InvalidOperationException ($"INTERNAL ERROR: Java type '{javaTypeName}' not found in cache");
+				}
 
-				moduleData.Entries.Add (new StructureInstance<TypeMapModuleEntry> (map_entry));
-			}
+				var key = new StructureInstance<TypeMapJava> (javaType);
+				int idx = javaMap.BinarySearch (key, javaNameHashComparer);
+				if (idx < 0) {
+					throw new InvalidOperationException ($"Could not map entry '{javaTypeName}' to array index");
+				}
 
-			allModulesData.Add (moduleData);
+				return (uint)idx;
+			}
 		}
 
-		protected override void MapStructures (LlvmIrGenerator generator)
+		// Generate hashes for all Java type names, then sort javaMap on the name hash.  This has to be done in the writing phase because hashes
+		// will depend on architecture (or, actually, on its bitness) and may differ between architectures (they will be the same for all architectures
+		// with the same bitness)
+		(List<ModuleMapData> allMapModulesData, List<ulong> javaMapHashes) PrepareMapsForWriting (LlvmIrGenerator generator)
 		{
-			generator.MapStructure<MonoImage> ();
-			typeMapJavaStructureInfo = generator.MapStructure<TypeMapJava> ();
-			typeMapModuleStructureInfo = generator.MapStructure<TypeMapModule> ();
-			typeMapModuleEntryStructureInfo = generator.MapStructure<TypeMapModuleEntry> ();
+			bool is64Bit = generator.Is64Bit;
+
+			// Generate Java type name hashes...
+			for (int i = 0; i < javaMap.Count; i++) {
+				TypeMapJava entry = javaMap[i].Obj;
+				entry.JavaNameHash = HashName (entry.JavaName);
+			}
+
+			// ...sort them...
+			javaMap.Sort ((StructureInstance<TypeMapJava> a, StructureInstance<TypeMapJava> b) => a.Obj.JavaNameHash.CompareTo (b.Obj.JavaNameHash));
+
+			var allMapModulesData = new List<ModuleMapData> ();
+
+			// ...and match managed types to Java...
+			foreach (StructureInstance<TypeMapModule> moduleInstance in mapModules) {
+				TypeMapModule module = moduleInstance.Obj;
+				PrepareMapModuleData (module.MapSymbolName, module.Data.Types, allMapModulesData);
+				if (module.Data.DuplicateTypes.Count > 0) {
+					PrepareMapModuleData (module.DuplicateMapSymbolName, module.Data.DuplicateTypes.Values, allMapModulesData);
+				}
+			}
+
+			var javaMapHashes = new List<ulong> (javaMap.Count);
```

Comment:
Does this need to be a `List<ulong>`, and not e.g. a `HashSet<ulong>`?  (Is ordering important?)

I think for "defense in depth" we should check the "impossible" of duplicate hashes.  Using a `HashSet<ulong>` would allow us to do this by checking the [`HashSet<T>.Add(T)`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.add?view=net-6.0) return value.  Otherwise we'd be looking at `List<T>.Contains(T)` calls to verify that there were no duplicates.

Reply:
It has to be a list, we must sort it before writing because it's searched with binary search at the run time. Duplicate hashes are extremely unlikely, so I decided not to search for them, but I can add a check if you insist :)